File: BinarySearch/FindSmallestLetter.py
from typing import List

class Solution:
    def nextGreatestLetter(self, letters: List[str], target: str) -> str:
        l, r = 0, len(letters)
        while l < r:
            mid = (l + r) // 2
            if letters[mid] <= target:
                l = mid + 1
            else:
                r = mid
        return letters[r] if r < len(letters) else letters[0]


let = ['c', 'f', 'j']
tar = 'c'
ss = Solution()
print(ss.nextGreatestLetter(let, tar))

File: BinarySearch/FindSmallestLetter.cpp
#include <iostream>
#include <vector>

using std::vector;

class Solution {
public:
    char nextGreatestLetter(vector<char>& letters, char target) {
        int l = 0, r = letters.size();
        while (l < r) {
            int mid = (l + r) / 2;
            if (letters[mid] <= target)
                l = mid + 1;
            else
                r = mid;
        }
        return r < letters.size() ? letters[r] : letters[0];
    }
};

int main()
{
    Solution s;
    char target = 'a';
    vector<char> letters {'c', 'f', 'j'};
    char c = s.nextGreatestLetter(letters, target);
    std::cout << c << "\n";
    return 0;
}
